workshops/evals/evals.ipynb

```diff
@@ -0,0 +1,757 @@
+                "# The model solution prints the message history, the score and the explanation of the score for each sample.\n",
+                "# Hide: hard\n",
+                "def print_info_from_sample(sample: EvalSample):\n",
+                "    print(\"---MESSAGE HISTORY---\")\n",
+                "    for message in sample.messages:\n",
+                "        if isinstance(message, ChatMessageUser):\n",
+                "            print(f\"User: {message.content}\" + \"\\n\")\n",
+                "        elif isinstance(message, ChatMessageAssistant):\n",
+                "            print(f\"Assistant: {message.content}\" + \"\\n\")\n",
+                "\n",
+                "    print(\"--SCORER INFORMATION---\")\n",
+                "    value = sample.scores[\"model_graded_fact\"].value\n",
+                "    print(f\"Score: {'correct' if value=='C' else 'incorrect'}\")\n",
+                "    print(f\"Explanation: {sample.scores['model_graded_fact'].explanation}\")\n",
+                "\n",
+                "\n",
+                "def print_info_from_logs(logs: list[EvalLog]):\n",
+                "    for i, sample in enumerate(logs[0].samples):\n",
+                "        print(f\"===START OF SAMPLE {i}===\")\n",
+                "        print_info_from_sample(sample)\n",
+                "        print(f\"===END OF SAMPLE {i}===\")\n",
+                "\n",
+                "\n",
+                "# Hide: none"
+            ]
+        },
+        {
+            "cell_type": "code",
+            "execution_count": null,
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "print_info_from_logs(logs)"
+            ]
+        },
+        {
+            "cell_type": "markdown",
+            "metadata": {},
+            "source": [
+                "By skimming through the message history and scorer's explanations, hopefully we can see how the different pieces of task have come together:\n",
+                "\n",
+                "Some questions to check your understanding:\n",
+                "- Which part(s) of the message history correspond to the input string from a sample?\n",
+                "- Where is the target string from a sample used?\n",
+                "- Which part(s) of the message history were affected/created by `chain_of_thought`? By `generate()`? By `self_critique`?\n",
+                "- Earlier we said that `self_critique` consists of two parts with two separate prompt templates.\n",
+                "Only one of the prompt templates actually appears in the message history.\n",
+                "What do you think is going on inside `self_critique` to explain this?"
+            ]
+        },
+        {
+            "cell_type": "markdown",
+            "metadata": {},
+            "source": [
+                "### OPTIONAL EXERCISE:\n",
```

Comment:
what questions do you anticipate getting about this? (i have still moved to details section to declutter a bit)

Reply:
The main questions I expect are: What's the answer? How can I figure that out?
In general it's good to have the solutions somewhere, so that people running the workshop in the future don't have to do the exercise themselves

Reply:
Oh. I don't know the answer myself... This is a cheeky way for me to get someone to check my own code. I think it is a bug in inspect but I am not sure. I dont rememebr this happening the first time I tried the basic example a few months ago

I will update the phrasing to make it clearer that this is a favour i am asking, rather than just an exercise.